File: DotNET/Examples/InventoryTwo/Observer/Subject.cs
using System;
using System.Collections;

using Composestar.RuntimeCore.FLIRT.Message;

namespace InventoryTwo
{
	/// <summary>
	/// Internal for subject filter module.
	/// </summary>
	public class Subject
	{
		private IList observers;

		public Subject()
		{
			observers = new ArrayList();
		}

		public void Attach(IObserver o)
		{
			observers.Add(o);
		}

		public void Detach(IObserver o)
		{
			observers.Remove(o);
		}

		public bool Observed()
		{
			return (observers.Count > 0);
		}

		public void Notify(ReifiedMessage rm)
		{
			rm.proceed();
			foreach (IObserver o in observers)
			{
				o.Update(this);
			}
		}
	}
}

File: Java/Examples/BookShelf/src/observer/Subject.java
package observer;

import java.util.ArrayList;
import java.util.List;

import Composestar.Java.FLIRT.Env.ReifiedMessage;

public class Subject {
    protected List<IObserver> observers;

    public Subject() {
	observers = new ArrayList<IObserver>();
    }

    public void attach(IObserver observer) {
	observers.add(observer);
    }

    public void detach(IObserver observer) {
	observers.remove(observer);
    }

    public void notify(ReifiedMessage msg) {
	Object subject = msg.getTarget();
	msg.proceed();
	for (IObserver obs : observers) {
	    obs.subjectChanged(subject);
	}
    }
}
